examples/viewer.py

```diff
@@ -0,0 +1,1489 @@
+        No-op for kinematic mode.
+        """
+
+    def updatePivot(self, pos: mn.Vector3) -> None:
+        pass
+
+    def update_frame(self, frame: mn.Matrix3x3) -> None:
+        pass
+
+    def _update_object_transform(self) -> None:
+        """
+        Apply the transform to the grabbed object.
+        """
+        obj = None
+        rom = self.simulator.get_rigid_object_manager()
+        aom = self.simulator.get_articulated_object_manager()
+        if rom.get_library_has_id(self.settings.object_id_a):
+            obj = rom.get_object_by_id(self.settings.object_id_a)
+        else:
+            # must be an ao
+            obj = aom.get_object_by_id(self.settings.object_id_a)
+        obj.transformation = (
+            self.global_mouse_grabber_transform @ self.relative_transformation
+        )
+
+    def update_transform(self, transform: mn.Matrix4) -> None:
+        self.global_mouse_grabber_transform = transform
+        self._update_object_transform()
+
+    def rotate_local_frame_by_global_angle_axis(
+        self, axis: mn.Vector3, angle: mn.Rad
+    ) -> None:
+        """rotate the object's local constraint frame with a global angle axis input."""
+        local_axis = self.relative_transformation.transform_vector(axis)
+        R = mn.Matrix4.rotation(angle, local_axis.normalized())
+        new_rotation = R.rotation().__matmul__(
+            self.relative_transformation.rotation()
+        )
+        self.relative_transformation = mn.Matrix4.from_(
+            new_rotation, self.relative_transformation.translation
+        )
+        self._update_object_transform()
+
+
+class Timer:
+    """
+    Timer class used to keep track of time between buffer swaps
+    and guide the display frame rate.
+    """
+
+    start_time = 0.0
+    prev_frame_time = 0.0
+    prev_frame_duration = 0.0
+    running = False
+
+    @staticmethod
+    def start() -> None:
+        """
+        Starts timer and resets previous frame time to the start time.
+        """
+        Timer.running = True
+        Timer.start_time = time.time()
+        Timer.prev_frame_time = Timer.start_time
+        Timer.prev_frame_duration = 0.0
+
+    @staticmethod
+    def stop() -> None:
+        """
+        Stops timer and erases any previous time data, resetting the timer.
+        """
+        Timer.running = False
+        Timer.start_time = 0.0
+        Timer.prev_frame_time = 0.0
+        Timer.prev_frame_duration = 0.0
+
+    @staticmethod
+    def next_frame() -> None:
+        """
+        Records previous frame duration and updates the previous frame timestamp
+        to the current time. If the timer is not currently running, perform nothing.
+        """
+        if not Timer.running:
+            return
+        Timer.prev_frame_duration = time.time() - Timer.prev_frame_time
+        Timer.prev_frame_time = time.time()
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser()
+
```

Comment:
could you add a comment on how to call this with a new dataset?

Reply:
Probably don't want to use the app on a new dataset because there are no kinematic objects unless we load an episode. This example app won't really be interesting in that case.